appscale/tools/node_layout.py

```diff
@@ -367,202 +317,259 @@ def is_valid_simple_format(self):
+          nodes[1].add_taskqueue_role(True)
+        else:
+          nodes[0].add_taskqueue_role(True)
+      if role == 'login':
+        login_found = True
+      elif role == 'login' and login_found:
+        self.invalid("Only one login is allowed")
 
-    if self.add_to_existing:
-      self.nodes = nodes
-      return self.valid()
+      # Update dictionary containing role counts.
+      role_count.update({(role, role_count.get(role, 0) + len(nodes))})
+      # Update the node_hash with the modified nodes.
+      node_hash.update({(node.public_ip, node) for node in nodes})
 
-    master_nodes = []
-    for node in nodes:
-      if node.is_role('shadow'):
-        master_nodes.append(node)
+      # Update node_count.
+      node_count += len(nodes)
 
-    # need exactly one master
-    if len(master_nodes) == 0:
-      return self.invalid("No master was specified")
-    elif len(master_nodes) > 1:
-      return self.invalid("Only one master is allowed")
+    # Dont need the hash any more, make a nodes list
+    nodes = node_hash.values()
 
-    master_node = master_nodes[0]
+    self.distribute_unassigned_roles(nodes, role_count, master_node)
 
-    login_nodes = []
-    for node in nodes:
-      if node.is_role('login'):
-        login_nodes.append(node)
+    if self.infrastructure in InfrastructureAgentFactory.VALID_AGENTS:
+      if not self.min_vms:
+        self.min_vms = len(nodes)
+      if not self.max_vms:
+        self.max_vms = len(nodes)
 
-    # If a login node was not specified, make the master into the login node
-    if not login_nodes:
-      master_node.add_role('login')
+    rep = self.is_database_replication_valid(nodes)
+    if not rep:
+      return rep
 
-    # by this point, somebody has a login role, so now's the time to see if we
-    # need to override their ip address with --login_host
-    if self.login_host is not None:
+    self.nodes = nodes
+
+    return True
+
+  def is_valid_advanced_format(self):
+    """Checks to see if this NodeLayout represents an acceptable (new) advanced
+    deployment strategy, and if so, constructs self.nodes from it.
+    
+    Returns:
+      True if the deployment strategy is valid.
+    Raises:
+      BadConfigurationException with reason if the deployment strategy is not 
+        valid.
+    """
+    # Keep track of whether the deployment is valid while going through.
+    node_hash = {}
+    role_count = {
+      'appengine': 0,
+      'master': 0,
+      'memcache': 0,
+      'taskqueue': 0,
+      'zookeeper': 0,
+      'login': 0
+    }
+    node_count = 1
+    db_master_created = False
+    tq_master_created = False
+    master_node = None
+    login_found = False
+    # Loop through the list of "node sets", which are grouped by role.
+    for node_set in self.input_yaml:
+      # If the key nodes is mapped to an integer it should be a cloud
+      # deployment.
+      cloud = True if isinstance(node_set.get('nodes'), int) else False
```

Comment:
`cloud = isinstance(node_set.get('nodes'), int)`. I would reconsider the name since It can be a cloud deployment without nodes being an integer.